File: leetcode/Python/sum-of-distances-in-tree.py
class Solution:
    def __init__(self):
        self.node_count = None
        self.sol = None
        self.graph = None
    def postorder(self, node, parent_node):
        '''
        calculate node_nums and dist value of child node
        '''
        self.node_count[node]=1
        for child_node in self.graph[node]:
            if child_node == parent_node:
                continue
            self.postorder(child_node,node)
            self.node_count[node]+=self.node_count[child_node]
            self.sol[node]+=self.sol[child_node]+self.node_count[child_node]
    def preorder(self, node, parent_node):
        '''
        calculate final dist val while traversiong node using (precalculated dist value of child_node)
        '''
        for child_node in self.graph[node]:
            if child_node == parent_node:
                continue
            self.sol[child_node]=self.sol[node]+len(self.graph)-2*self.node_count[child_node]
            self.preorder(child_node,node)
    def sumOfDistancesInTree(self, n: int, edges: List[List[int]]) -> List[int]:
        # 1. init variable
        self.node_count = [0]*n
        self.sol = [0]*n
        self.graph = [list() for i in range(n)]
        # 2. make graph
        for e in edges:
            self.graph[e[0]].append(e[1])
            self.graph[e[1]].append(e[0])
        # 3. postorder (to get dist, node_count of child)
        self.postorder(0,0)
        # 4. preorder (to get final dist value using precalculated dist of child)
        self.preorder(0,0)
        return self.sol
    

File: leetcode/C++/sum-of-distances-in-tree.cpp
class Solution {
public:
    vector<vector<int>> graph;
    vector<int> sol,node_count;
    // calculate node_nums and dist value of child node
    void postorder(int node, int parent_node) {
        node_count[node] = 1; // self node count
        for(int &child_node : graph[node]){
            if(child_node==parent_node) continue;
            postorder(child_node,node);
            node_count[node]+=node_count[child_node];
            sol[node]+=sol[child_node]+node_count[child_node];
        }
    }
    // calculate final dist val while traversiong node using (precalculated dist value of child_node)
    void preorder(int node, int parent_node){
        for(int &child_node : graph[node]){
            if(child_node==parent_node) continue;
            sol[child_node]=sol[node]+graph.size()-2*node_count[child_node];
            preorder(child_node, node);
        }
    }
    vector<int> sumOfDistancesInTree(int n, vector<vector<int>>& edges) {
        //1. init variable
        graph = vector<vector<int>>(n);
        sol = vector<int>(n,0);
        node_count = vector<int>(n,0);
        //2. make graph
        for(vector<int> &edge : edges){
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);
        }
        //3. postorder (to get dist, node_count of child)
        postorder(0,0);
        //4. preorder (to get final dist value using precalculated dist of child)
        preorder(0,0);
        return sol;
    }
};